File: java/src/main/java/io/binac/leetcode/SingleNumberII.java
package io.binac.leetcode;

/**
 * Given a non-empty array of integers, every element appears three times except for one, which appears exactly once. Find that single one.
 * <p>
 * <p>Note:
 * <p>
 * <p>Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: [2,2,3,2]
 *     Output: 3
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: [0,1,0,1,0,1,99]
 *     Output: 99
 * </blockquote></pre>
 */
public class SingleNumberII {
    public static class Solution1 {
        public int singleNumber(int[] nums) {
            int one = 0, two = 0;
            for (int num : nums) {
                one = (one ^ num) & ~two;
                two = (two ^ num) & ~one;
            }
            return one;
        }
    }
}

File: python/leetcodepy/single_number_ii.py
"""
Given a non-empty array of integers, every element appears three times except for one, which appears exactly once. Find that single one.

Note:

Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?

Example 1:

Input: [2,2,3,2]
Output: 3

Example 2:

Input: [0,1,0,1,0,1,99]
Output: 99
"""
from typing import List


class Solution1:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        one, two = 0, 0
        for num in nums:
            one = ~two & (one ^ num)
            two = ~one & (two ^ num)
        if two != 0:
            raise ValueError
        return one
